src/subdomains/generic/user/services/webhook/webhook.service.ts

```diff
@@ -14,100 +13,114 @@ import { WebhookRepository } from './webhook.repository';
+    const payload = WebhookDataMapper.mapAccountMergeData(master, slave);
+
+    await this.sendWebhooks(WebhookType.ACCOUNT_MERGE, payload, slave, []);
+  }
+
+  // --- PAYMENT WEBHOOKS --- //
   async fiatCryptoUpdate(user: User, payment: BuyCryptoExtended): Promise<void> {
-    await this.create({
-      user,
-      data: JSON.stringify(WebhookDataMapper.mapFiatCryptoData(payment)),
-      type: WebhookType.PAYMENT,
-    });
+    const payload = WebhookDataMapper.mapFiatCryptoData(payment);
+
+    await this.sendWebhooks(WebhookType.PAYMENT, payload, user.userData, [user]);
   }
 
   async cryptoCryptoUpdate(user: User, payment: BuyCryptoExtended): Promise<void> {
-    await this.create({
-      user,
-      data: JSON.stringify(WebhookDataMapper.mapCryptoCryptoData(payment)),
-      type: WebhookType.PAYMENT,
-    });
+    const payload = WebhookDataMapper.mapCryptoCryptoData(payment);
+
+    await this.sendWebhooks(WebhookType.PAYMENT, payload, user.userData, [user]);
   }
 
   async cryptoFiatUpdate(user: User, payment: BuyFiatExtended): Promise<void> {
-    await this.create({
-      user,
-      data: JSON.stringify(WebhookDataMapper.mapCryptoFiatData(payment)),
-      type: WebhookType.PAYMENT,
-    });
+    const payload = WebhookDataMapper.mapCryptoFiatData(payment);
+
+    await this.sendWebhooks(WebhookType.PAYMENT, payload, user.userData, [user]);
   }
 
   async fiatFiatUpdate(user: User, payment: BuyFiatExtended): Promise<void> {
-    await this.create({
-      user,
-      data: JSON.stringify(WebhookDataMapper.mapFiatFiatData(payment)),
-      type: WebhookType.PAYMENT,
-    });
+    const payload = WebhookDataMapper.mapFiatFiatData(payment);
+
+    await this.sendWebhooks(WebhookType.PAYMENT, payload, user.userData, [user]);
   }
 
   // --- HELPER METHODS --- //
+  private async sendWebhooks(
+    type: WebhookType,
+    payload: object,
+    userData: UserData,
+    users: User[],
+    reason?: string,
+  ): Promise<void> {
+    const data = JSON.stringify(payload);
+
+    // user webhooks
+    const webhooks: CreateWebhookInput[] = users
+      .filter((user) => user.wallet?.isValidForWebhook(type, false))
+      .map((user) => ({ type, data, reason, userData, user, wallet: user.wallet }));
 
-  private async create(dto: CreateWebhookInput): Promise<Webhook | undefined> {
-    if (!dto.user.wallet)
-      dto.user = await this.userRepo.findOne({ where: { id: dto.user.id }, relations: { wallet: true } });
-    if (!dto.user.wallet.apiUrl) return;
+    // user data webhooks
+    const additionalClients = userData.kycClientList.filter((w) => !webhooks.some(({ wallet }) => wallet.id === w));
+    for (const walletId of additionalClients) {
+      const wallet = await this.walletService.getByIdOrName(walletId);
+      if (wallet?.isValidForWebhook(type, true)) webhooks.push({ type, data, reason, userData, wallet });
+    }
+
+    for (const client of webhooks) {
+      await this.createAndSendWebhook(client);
+    }
```

Comment:
Do we need to load the wallet of the user, in case it's not loaded?

Reply:
Yeah I think so, or?

Reply:
Just pushed